BLSQ/iaso
# Polio 619 refactor budget front end

Refactor Front-end of the budget tool to accomodate the use of the workflow engine

## Self proof reading checklist

- [ ] Are my typescript files well typed - half-half
- [ ] Are there enough tests - Not in the front
## Changes

###  Features
- Add step button + modal has been replaced by buttons generated dynamically according to the workflow state
- The Map in Budget details has been removed to give more space to the Table, which allowed to remove the `BudgetFiles`modal in the desktop version: the files and links are shown directly in the table
- Budget Steps are not editable anymore
- The notion of deletion/restoration of events has been replaced by show/hide, which don't open a modal anymore.
- Filters by team and recipient have been removed as they don't make sense anymore in the context of the steps being constrained by the workflow.
- When adding links, users can now choose an alias, that will be displayed in the table
- Step creation  buttons can have 3 colors: "primary", "red" or "green". These are determined in the workflow

### In the code

- Data for the Budget Table is received from a single endpoint from the backend. No values are computed in the front-end (as opposed to previously where e.g. status was computed from all events for the campaign)
- All calls to `/api/polio/campaign` have been removed
- `CreateEditBudgetEvent`has been replaced with `CreateBudgetStep` and `CreateOverrideStep`
- Table and Card layouts for BudgetDetails have been split and moved to their own component
- Budget status and step names come directly from the backend and are not translated in the front-end
- Added `AddMultipleLinks` and `NamedLink`components for managing links' aliases
- Most calls to `api/profiles` have been removed since the backend handles  the necessary checks
- links are now treated as a list of objects with an `alias` and a `url`
- likewise have a `filename` that we can display in the table. 
- Step creation button colors are handled with classNames, which explains why the are constrained to predetermined values


## How to test

- Get the workflow config file from [Google Drive](https://drive.google.com/drive/u/0/folders/1ua4yzyFOGayweQVNKP0hyBNDKhQrPlx1)
- Go to Django admin -> workflow models and paste the json from the file there
- To test, you will need need the following teams with the corresponding team ids: 
    - 1: RRT
    - 2: ORPG Ops
    - 3: ORPG wider group
    - 4: UNICEF
    - 5: WHO
    - 6: MoH (stands for Ministry of Health)
    - 7: GPEI coordinator
 - If the team ids are not free, you will need to change the workflow config with the ids you will use. If you're using a polio DB it should work since no teams were created there yet
 - You will then need a user for each team. Ideally you need to make separate users, otherwise you will not be able to test that the actions are restricted based on the team
 - Once all this is set up, you can test the workflow by selecting a campaign and starting to create steps. The "scenario" of the workflow is in [Whimsical](https://whimsical.com/latest-version-budget-workflows-RkEXevU68FbkjJScFzJU2w)

## Print screen / video

https://user-images.githubusercontent.com/38907762/194546197-392f87e7-2423-41b3-987d-374ca4a7fc6f.mov


## Notes

⚠️ Includes [POLIO-617](https://github.com/BLSQ/iaso/tree/POLIO-617_BWE_backend) changes (backend)

It's a big PR since it incorporates current work of @olethanh on the backend, so:
@niconoe Can you please have a look at the backend part
@beygorghor Can you check the front

Both are quite big in their own right, though @beygorghor  you'll see that a lot of the front-end changes are renaming stuff or replacing types

Known issues:

- Modal fields for each step are not yet dynamic ("displayed fields")
- Backend for override feature is not ready
- Display bug with next steps when there are several possible options
